pkg/controller/pravegacluster/upgrade.go

```diff
@@ -404,6 +405,178 @@ func (r *ReconcilePravegaCluster) syncSegmentStoreVersion(p *pravegav1alpha1.Pra
+	name = util.StatefulSetNameForSegmentstoreBelow07(p.Name)
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: name, Namespace: p.Namespace}, stsBelow07)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return false
+		}
+		log.Printf("failed to get PravegaCluster: %v", err)
+		return false
+	}
+	//true will only be return if both sts are present
+	return true
+}
+
+//this function calls the syncstoreversion for upgrade and rollback operation
+func (r *ReconcilePravegaCluster) syncStoreVersion(p *pravegav1alpha1.PravegaCluster) (synced bool, err error) {
+	//in case of a upgrade from a version below 07 to a version above 07 syncSegmentStoreVersionTo07() is called and the util.IsClusterUpgradingTo07(p)  is true in this case
+	//in case of a rollback when upgrade fails from a version below 07 to a version above 07 syncSegmentStoreVersionTo07() is called and the r.IsClusterRollbackingFrom07(p) is true in this case
+	if util.IsClusterUpgradingTo07(p) || r.IsClusterRollbackingFrom07(p) {
+		return r.syncSegmentStoreVersionTo07(p)
+	}
+	//for all other cases of upgrades and rollback this function is called
+	return r.syncSegmentStoreVersion(p)
+}
+
+//this fuction is called when we want to upgrade ss from a version below 07 to a version above 07  as well as when the such upgrades fail then for doing rollback also it's called
+func (r *ReconcilePravegaCluster) syncSegmentStoreVersionTo07(p *pravegav1alpha1.PravegaCluster) (synced bool, err error) {
+	p.Status.UpdateProgress(pravegav1alpha1.UpdatingSegmentstoreReason, "0")
+	newsts := pravega.MakeSegmentStoreStatefulSet(p)
+	controllerutil.SetControllerReference(p, newsts, r.scheme)
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: newsts.Name, Namespace: p.Namespace}, newsts)
+	//this check is to see if the newsts is present or not if it's not present it will be created here
+	if err != nil && errors.IsNotFound(err) {
+		*newsts.Spec.Replicas = 0
+		err = r.client.Create(context.TODO(), newsts)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	//here we are getting the name of the oldsts based on either we are upgrading or doing rollback
+	var oldstsName string = ""
+	oldsts := &appsv1.StatefulSet{}
+	if util.IsClusterUpgradingTo07(p) {
+		oldstsName = util.StatefulSetNameForSegmentstoreBelow07(p.Name)
+	} else {
+		oldstsName = util.StatefulSetNameForSegmentstoreAbove07(p.Name)
+	}
+
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: oldstsName, Namespace: p.Namespace}, oldsts)
+	//this check is to see if the old sts is present or not
+	if err != nil && errors.IsNotFound(err) {
+		//this is the condition where we are checking if the new ready ss pods are equal to the ss replicas thus meaning upgrade is completed
+		if newsts.Status.ReadyReplicas == p.Spec.Pravega.SegmentStoreReplicas {
+			return true, nil
+		}
+	}
+
+	//TO detect upgade/rolback faiure
+	if oldsts.Status.ReadyReplicas+newsts.Status.ReadyReplicas != p.Spec.Pravega.SegmentStoreReplicas {
```

Comment:
!= check should be changed to < as discussed.

Reply:
i have changed it